### PR TITLE
feat: return relevant stack trace as error from ivm

### PR DESCRIPTION
**Pull request by dparveez-ruddr**
Return relevant stack trace as error from ivm. Some basic transformation/extraction done on stack trace.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/util/ivmFactory.js [hunk 1]
```diff
@@ -246,6 +249,10 @@ async function createIvm(code, libraryVersionIds, versionId, secrets, testMode)
     }
   });
 
+  await jail.set('extractStackTrace', function(trace, stringLiterals) {
+    return extractStackTraceUptoLastSubstringMatch(trace, stringLiterals);
```
Comment (Jayachand) [inline src/util/ivmFactory.js:253]: add this for v0 transformation too - customTransformer.js

**Review by Jayachand — DISMISSED** [review]

File: test/__tests__/data/user_transformation_async_output_with_errors.json [hunk 2]
```diff
@@ -41,11 +41,11 @@
     "metadata": {}
   },
   {
-    "error": "Error: Non-identify found",
+    "error": "Error: Non-identify found\n    at rejectNonIdentifies (base transformation:7:15)\n    at transformEvent (base transformation:10:28)",
     "metadata": {}
   },
   {
-    "error": "Error: Non-identify found",
+    "error": "Error: Non-identify found\n    at rejectNonIdentifies (base transformation:7:15)\n    at transformEvent (base transformation:10:28)",
```
Comment (Jayachand) [inline test/__tests__/data/user_transformation_async_output_with_errors.json:48]: write a test suite handling these cases:
```
transformEvent -> error inside (testMode as true)
transformEvent with library -> error in library
transformEvent -> throw error explicitly
transformBatch -> random error in code
transformBatch with library -> error in library
transformBatch -> throw error explicitly
```

File: src/util/utils.js [hunk 3]
```diff
@@ -159,6 +159,22 @@ function logProcessInfo() {
   logger.error(`Process info: `, util.inspect(processInfo(), false, null, true));
 }
 
+// stringLiterals expected to be an array of strings. A line in trace should contain
+// atleast one string in the stringLiterals array for lastMatchingIndex to be updated appropriately.
+const extractStackTraceUptoLastSubstringMatch = (trace, stringLiterals) => {
+  let lastRelevantIndex = 0;
+
+  const traceLines = trace.split('\n');
```
Comment (koladilip) [inline src/util/utils.js:167]: can't we look in reverse order and break on first match?

File: test/__tests__/user_transformation_errors.test.js [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+jest.mock("node-fetch");
+const fetch = require("node-fetch", () => jest.fn());
+
+const {
+  userTransformHandler
+} = require("../../src/util/customTransformer");
+
+const events = [
+  {
+    message: {
```
Comment (Jayachand) [inline test/__tests__/user_transformation_errors.test.js:12]: events must have messageId

File: test/__tests__/user_transformation_errors.test.js [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+        events,
+        versionId,
+        [],
+        trRevCode,
+        true,
+      );
+            
+      expect(result.transformedEvents[0].error).toEqual(
+        "ReferenceError: events is not defined\n    at transformEvent (base transformation:2:2)"
+      )
```
Comment (Jayachand) [inline test/__tests__/user_transformation_errors.test.js:48]:
```suggestion
      expect(result.transformedEvents.length).toBe(3);
      result.transformedEvents.forEach(ev => {expect(ev.error).toEqual(
        "ReferenceError: events is not defined\n    at transformEvent (base transformation:2:2)"
      )})
```
nit: prefer this way. validating output events count as well as each event

**Review by Jayachand — APPROVED** [review]

**Pull request merged**